### PR TITLE
Parallelizing the bucket brigade qRAM with CCZs

### PR DESCRIPTION
**Pull request by glassnotes**
This PR contains:
 - implementation (and tests) of bucket-brigade bit query with heavy T gate parallelization using CCZs, according to the work of Alexandru et al.; covers both single- and multi-output bit cases
 - updates to the primer with description and circuit diagrams for the above
 - updates to the primer detailing how to implement the next step (a phase query)

(Note: this includes all the content of glassnotes/ccz-bucket-brigade, I merged it in here before updating the primer)

### CODE REVIEW

File: src/BucketBrigadeCCZ.qs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+///////////////////////////////////////////////////////////////////////////
+    
+    /// # Summary
+    /// Writes a single bit of data to the memory.
+    /// # Input
+    /// ## memoryRegister
+    /// Register that represents the memory you are writing to.
+    /// ## dataValue
+    /// The tuple of (address, data) that you want written to the memory.
+    operation BucketBrigadeCCZWrite(
```
Comment (RolfHuisman) [inline src/BucketBrigadeCCZ.qs:51]: Shouldn't these be internal, since they are under the \\\internal ?

File: tests/CommonTest.qs [hunk 2]
```diff
@@ -52,8 +53,8 @@ namespace Tests {
     }
 
     // QRAM where only the last memory cell contains a 1
-    internal function LastCellFullQRAM(addressSize : Int) : MemoryBank {
-        return GeneratedMemoryBank([MemoryCell(2^addressSize - 1, [true])]);
+    internal function LastCellFullQRAM(addressBits : Int) : MemoryBank {
```
Comment (RolfHuisman) [inline tests/CommonTest.qs]: AddressBits with type int implies a bit that this is a mask or a value mask. Not that problematic, but given the rest uses the size word for this, might be better to keep it consistent in the file (unless you really mean a mask here).

File: src/Common.qs [hunk 3]
```diff
@@ -40,6 +40,8 @@ namespace Qram{
     /// # Summary
     /// Wrapper for registers that represent a quantum memory.
     newtype MemoryRegister = (Qubit[]);
+
```
Comment (RolfHuisman) [inline src/Common.qs:45]: Some lose enters.

File: docs/primer.tex [hunk 4]
```diff
@@ -459,8 +461,130 @@ \subsubsection{Resource estimation}
...
+\subsection{Improving $T$ depth with the CCZ gate}
+
+Looking at the circuits of \autoref{fig:bb-combined}, one notes that there are many instances of sequential Toffolis. 
+In \cite{DiMatteo2020}, a parallelized version of the circuit was proposed. 
+In the fanout, multiple copies of the address bits were made so the Toffolis could be performed in parallel. 
+The query portion was parallelized by giving each Toffoli a different target qubit, and then taking the parity of those bits using CNOT gates. 
+This works fine, and has the effect of reducing the $T$-depth from $O(2^n)$ to $O(n)$. 
+However, it comes at a significant cost: the number of auxiliary qubits that must be added is $O(2^n)$.
+
+\todo{These sections are rough, and will be re-written as I make my way through \cite{Alexandru2020}.}
```
Comment (RolfHuisman) [inline docs/primer.tex:473]: Its quite readable already actually.

File: docs/primer.tex [hunk 5]
```diff
@@ -459,8 +461,130 @@ \subsubsection{Resource estimation}
...
+
+\todo{Add discussion of circuit depth and complexity of `quantum fanout'.}
+
 \subsection{Phase query circuit model}
 
+In some situations, most notably in Grover's algorithm, one is concerned with retrieving the memory contents not in bit form, but as a phase. 
+Explicitly, 
+\begin{equation}
+ \ket{i} \rightarrow (-1)^{m_i} \ket{i}
+\end{equation}
```
Comment (RolfHuisman) [inline docs/primer.tex:569]: What do you mean with this remark ? Normally I'm familiar that the data is not by itself phased, only the shor algorithm itself makes use of it internally. Or is this a different use of grover in a different context?

File: src/BucketBrigade.qs [hunk 6]
```diff
@@ -94,7 +94,7 @@
     /// as a one-hot encoding.
     /// ## targetRegister
     /// The register that will have the memory value transferred to.
-    internal operation ReadoutMemory(
```
Comment (RolfHuisman) [inline src/BucketBrigade.qs:97]: This one needs to move to a utility place I think since its being used more broadly than before.

**Review by RolfHuisman — COMMENTED** [review]
Some comments while reading the changes.

File: src/BucketBrigadeCCZ.qs (hunk 1 above)
Comment (glassnotes) [reply src/BucketBrigadeCCZ.qs:51]: Oh nice catch! I'm not sure - some of the stuff I needed to make not internal in order to use them in the tests. Is there a way to like make them available for tests only but not for a user?

Comment (RolfHuisman) [reply src/BucketBrigadeCCZ.qs:51]: In C# and VB there is normally the attribute internalsvisibleto one can put one the assembly for that (makes the internals usable for a test library). Don't know if Q# supports that. @crazy4pi314 do you know ?

File: tests/CommonTest.qs (hunk 2 above)
Comment (glassnotes) [reply tests/CommonTest.qs]: Yup, that should be addressSize in line with the rest.

File: docs/primer.tex (hunk 4 above)
Comment (glassnotes) [reply docs/primer.tex:473]: Oh thanks! :)

File: docs/primer.tex (hunk 5 above)
Comment (glassnotes) [reply docs/primer.tex:569]: In Grover there is the oracle that adds a (-1) phase only to marked elements of the search space. The idea is to do something similar here, like add the phase only when the element is "marked" (1). @crazy4pi314 and I were discussing a bit this morning, we have to work out more details on what this means for the multi-bit case.

Comment (RolfHuisman) [reply docs/primer.tex:569]: Gotcha.

File: src/BucketBrigadeCCZ.qs (hunk 1 above)
Comment (crazy4pi314) [reply src/BucketBrigadeCCZ.qs:51]: Not that I know of, probably best we try and follow the style guide on where to use? https://docs.microsoft.com/en-us/quantum/resources/contributing/style-guide?tabs=guidance#private-or-internal-names

And yeah for here, I don't think we want users using `BucketBrigadeCCZWrite` directly do we?

As to resolve the testing problems, I'll maybe ping the Q# team on the MVP Teams to see if there is a recommended structure for making that work.

File: src/BucketBrigade.qs (hunk 6 above)
Comment (crazy4pi314) [reply src/BucketBrigade.qs:97]: I'll take care of that

File: src/BucketBrigadeCCZ.qs (hunk 1 above)
Comment (crazy4pi314) [reply src/BucketBrigadeCCZ.qs:51]: We may have to leave them exposed for now so tests pass, Imma add a big TODO to make this internal once we figure out how to solve the testing thing.

**Review by crazy4pi314 — APPROVED** [review]
Heh I added single comments, forgot to start "Review". Looks fab, we can work on how to do access modifiers for testing.

**Pull request merged**